### ISSUE TITLE
Sync scripts: allow a production build with single sync

### ISSUE THREAD

**Issue by noahtallen** [Automattic/wp-calypso#48831]
## Overview

We have a sync script for ETK and wpcom-block-editor which can rsync changes to your sandbox. This script should support a one-off mode, where you can build and then sync the files once.

Props to @jsnajdr for the request

## Context

n/a

**Comment by scinos** [comment]
Maybe there is an opportunity here to standardize the 'sync with sandbox' flow across the whole repo. AFAIK all apps under `apps/` use it.

**Comment by jsnajdr** [comment]
> Maybe there is an opportunity here to standardize the 'sync with sandbox' flow across the whole repo.

:+1: All `apps` do the same thing when syncing, except the source and target directories. And the implementation is currently inconsistent. `o2-blocks` don't have a sync task at all. `notifications` sync only once and don't support watching with `chokidar` after that. `wpcom-block-editor` has the opposite bug, not doing the initial sync and only watching for changes while running.

**Comment by noahtallen** [comment]
Yep. I wrote the sync script for ETK a year ago, and I think some folks have copied the implementation for the other apps. I completely agree that we should make it consistent for the whole repo. Do you think it makes sense to add it to the `calypso-build` package? If not, what other place makes more sense?

Also related: #48830

**Comment by scinos** [comment]
I'd start with a new private package, and once it is stable and we have find the right abstractions, maybe merge it with `calypso-build` if we think it can be consumed outside the repo.

I'll love to take this opportunity to take a step back and think about the devloop in a more generic way. Things like: what should be the developer & delivery process for "apps" whose main deliverable artifact is a diff in wpcom repo? what should be the main model, sync to sandbox, rely on `wp-env` or something else? What's the best way to "deploy" and test a remote branch (as opposed to a local checkout)? what does "release", "rollback", "hotfix"...  even mean for that kind of projects?

**Comment by noahtallen** [comment]
Definitely, those are all good questions to ask, and ones I'd love to resolve. Some thoughts:

- I don't think we can rely on `wp-env` to containerize much for simple site development. It works nicely for local testing, and closely approximates self-hosted or non-transferred Atomic sites, but testing on a WordPress.com sandbox is very important before releasing
- This post (pMz3w-crN-p2) gets at an idea for testing a remote branch on a wpcom sandbox (assuming one could also transfer a wp-calypso app branch into a wpcom git branch with some sort of tool)
- If we do need to get code onto a wordpress.com sandbox for testing, we basically have to do some sort of "local -> sandbox" sync, or "GitHub PR -> sandbox" sync. We actually have solutions for both of these, and the "local -> sandbox" sync is very fast and requires no extra setup on the sandbox side of things. (There is a dedicated "dev" folder for the plugin version which is loaded only on the sandbox and only if a dev version exists). The "GitHub PR -> sandbox" option is limited mostly by two things: firstly, the length of time for GitHub to finish building and uploading the code. This takes about 6 minutes, mostly from `yarn install`. Secondly, the length of time to download the phabricator patch to your sandbox, which is maybe 5 minutes or more. I think `arc patch` is very inefficient for large diffs. And the diff is large because every new version of the plugin is deployed to a new directory, to avoid having to worry about problems with atomicity in the deploy system (pMz3w-bxL-p2)

**Comment by scinos** [comment]
> The "GitHub PR -> sandbox" option is limited mostly by two things: firstly, the length of time for GitHub to finish building and uploading the code. This takes about 6 minutes.

Maybe we can use TeamCity? Looking at current unit test times, it can checkout a branch, install dependencies and build everything under `apps/` in less than 2:30 minutes. Not sure about the "uploading code" part, I'd need to understand where it is being uploaded.

**Comment by noahtallen** [comment]
That sounds awesome. The upload step (which posts the .zip file to mission control) is only a few seconds.

**Comment by noahtallen** [comment]
I imagine we can rely somewhat on cached `yarn` deps fro TeamCity? I have the cache set up in GitHub actions, but I think it invalidates for every new PR.

**Comment by noahtallen** [comment]
I think this should be fixed.
